framework/src/outputs/formatters/SONDefinitionFormatter.C

```diff
@@ -225,11 +233,29 @@ SONDefinitionFormatter::addBlock(const std::string & block_name,
 void
 SONDefinitionFormatter::addParameters(const JsonVal & params)
 {
+
   for (const auto & param_name : params.getMemberNames())
   {
 
     JsonVal param = params[param_name];
 
+    // lambda to calculate relative path from the current level to the document root
+    auto backtrack = [this]() -> std::string {
```

Comment:
Why close over the ``this`` pointer?  If you want to use a lambda, I'd rather you pass in level as an arg and not close over anything.  Or even better - why can't we just build the backtrack string incrementally inside the (enormous) for loop and not use a lambda at all?

Reply:
@rwcarlsen 

I completely agree that i should just pass in `_level` as an argument to `backtrack` instead of closing over `this`

I cannot build the backtrack string just inside this for loop because it is based on which `_level` we are at in the SON DOM and that `_level` is incremented and decremented in the `addBlock` method above.  This for loop just loops over the parameters of the current block.

I could, however, build the backtrack string throughout the entire code whenever `_level` is incremented or decremented.  I thought about that when writing this, but I ran some numbers and saw that we are doing 54,848 increments and 54,848 decrements of `_level` for a total of 109,696 level changes in the document where we would be doing string manipulation at all of those.

The only two places where backtrack is used or where we every need a relative path up to the root is in the `ChildAtLeastOne` rule and the `ExistsIn` rule.

The `ChildAtLeastOne` rule is for `required` parameters that must either exist at the current level or back up inside the document_root/GlobalParams block.  This rule is used 2,984 times currently.

The `ExistsIn` rule is for looking up where something must exists if should have been defined elsewhere.  This rule is used 3,615 time currently.

So, when I thought about doing 109,696 string manipulations every time we go up or down a level in the hierarchy vs. only building this string when needed (2,984 +3,615 = 6,599 times), it seemed better to go the latter route.

I can definitely change this lambda to just take the `_level` as a parameter though instead of using `this`